subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java

```diff
@@ -0,0 +1,340 @@
+        DependencyVerificationsXmlWriter.serialize(
+            verificationsBuilder.build(),
+            new FileOutputStream(verifFile)
+        );
+    }
+
+    private void maybeReadExistingFile(File verifFile) throws FileNotFoundException {
+        if (verifFile.exists()) {
+            LOGGER.info("Found dependency verification metadata file, updating");
+            DependencyVerificationsXmlReader.readFromXml(new FileInputStream(verifFile), verificationsBuilder);
+        }
+    }
+
+    private void writeEntriesSerially() {
+        entriesToBeWritten.stream()
+            .sorted()
+            .forEachOrdered(entry -> {
+                verificationsBuilder.addChecksum(entry.id, entry.checksumKind, entry.checksum);
+            });
+    }
+
+    private void computeHashsConcurrently(Gradle gradle) {
+        buildOperationExecutor.runAll(queue -> {
+            Set<Project> allprojects = gradle.getRootProject().getAllprojects();
+            for (Project project : allprojects) {
+                queue.add(new RunnableBuildOperation() {
+                    @Override
+                    public void run(BuildOperationContext context) {
+                        resolveAllConfigurationsAndForceDownload(project);
+                    }
+
+                    @Override
+                    public BuildOperationDescriptor.Builder description() {
+                        return BuildOperationDescriptor.displayName("Computing dependency verification metadata for " + project.getDisplayName());
+                    }
+                });
+            }
+        });
+        // We don't need this anymore and because we're going to sort
+        // elements to be added for reproducibility, we're freeing memory just in case
+        cachedChecksums.clear();
+    }
+
+    @Override
+    public void onArtifact(ModuleComponentArtifactIdentifier id, File file) {
+        for (String checksum : checksums) {
+            addChecksum(id, file, ChecksumKind.valueOf(checksum));
+        }
+    }
+
+    private void addChecksum(ModuleComponentArtifactIdentifier id, File file, ChecksumKind kind) {
+        ChecksumEntry entry = buildOperationExecutor.call(new CallableBuildOperation<ChecksumEntry>() {
+            @Override
+            public ChecksumEntry call(BuildOperationContext context) {
+                return new ChecksumEntry(id, file, kind, createHash(file, kind));
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Generate " + kind + " checksum for " + id);
+            }
+        });
+        synchronized (entriesToBeWritten) {
+            entriesToBeWritten.add(entry);
+        }
+    }
+
+    private String createHash(File file, ChecksumKind kind) {
+        return cachedChecksums.computeIfAbsent(new FileChecksum(file, kind), key -> HashUtil.createHash(file, kind.getAlgorithm()).asHexString());
```

Comment:
This cache feels like it is at the wrong level. In `entriesToBeWritten`, it should be impossible to have twice the same `ModuleComponentArtifactIdentifier` with different file in a single build, otherwise it means we are effectively getting two different files over the course of a single execution. Right?
If so, we should make it an error the moment it happens and thus before computing a checksum we could check whether we got an entry for a given `ModuleComponentArtifactIdentifier` inside a `Map` and thus skip altogether the checksum computation, allowing to get rid of this cache.
Did I miss something?

Reply:
This is just an optimization. While I _could_ have a map which key is the `MCAI`, the value would be a list of checksums. Because it happens concurrently (this writer is global to a build, multiple resolution can happen concurrently and the same file can be found in different configurations), I can decouple the computation of the hash (for a file) with the set of entries. So it avoids the creation of an intermediate list of checksums.

Reply:
Realized that detecting the corner case of a single build loading different versions of the same `ModuleComponentArtifactIdentifier` is pretty corner case and complicated by the fact that there will be a number of duplicate handling since in a multi project there is a high chance we see multiple times the same `ModuleComponentArtifactIdentifier` across projects or configurations.
Let's got with this then!